### PR TITLE
feat: middlewares and exception handlers in LitServer.

### PR DESCRIPTION
**Pull request by lorenzomassimiani**
<details>
  <summary><b>Before submitting</b></summary>

- [x] Was this discussed/agreed via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/main/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

</details>

## User perspective

As a user I need to enrich the litserve server with fastapi middleware, for example to update API metrics.

## What does this PR do?

Added `middlewares` params to LitServer, that are assigned to the fastapi app in the `setup_server` method (in which I also moved the assignation of `MaxSizeMiddleware` and `GZipMiddleware` middlewares).  

`middlewares` is a list of 2-sized tuples, the first argument is the Middleware class, the second is a dict that contains possible kwargs.  

Be free to give me suggestions for improving the code and I am available to answer all questions.  

Thanks for your work :smiley: .

### CODE REVIEW

**Comment by williamFalcon** [comment]
awesome PR @lorenzomassimiani! can you add the motivation from the user perspective at the very top?
"As a user we need X because of Y"

File: src/litserve/server.py [hunk 1]
```diff
@@ -422,6 +424,8 @@ def __init__(
         stream: bool = False,
         spec: Optional[LitSpec] = None,
         max_payload_size=None,
+        middlewares: Optional[list[tuple[type[_MiddlewareClass[P]]], dict]] = None,
+        exc_handlers: Optional[list[tuple[int | type[Exception], ExceptionHandler]]] = None,
```
Comment (aniketmaurya) [inline src/litserve/server.py:428]: please keep the scope minimal and only implement middlewares in this PR.

Comment (aniketmaurya) [inline src/litserve/server.py:428]:
```suggestion
```

File: src/litserve/server.py [hunk 2]
```diff
@@ -422,6 +424,8 @@ def __init__(
         stream: bool = False,
         spec: Optional[LitSpec] = None,
         max_payload_size=None,
+        middlewares: Optional[list[tuple[type[_MiddlewareClass[P]]], dict]] = None,
```
Comment (aniketmaurya) [inline src/litserve/server.py:427]: `|` operator is not supported in Python 3.8 I think so you might need to use `typing.Union`.

File: src/litserve/server.py [hunk 3]
```diff
@@ -454,17 +458,15 @@ def __init__(
         self.app.response_queue_id = None
         self.response_queue_id = None
         self.response_buffer = {}
-        # gzip does not play nicely with streaming, see https://github.com/tiangolo/fastapi/discussions/8448
```
Comment (aniketmaurya) [inline src/litserve/server.py:457]: you can append these middleswares to `self.middlewares` and add it later below in `setup_server` method

File: src/litserve/server.py [hunk 4]
```diff
@@ -668,7 +670,20 @@ async def stream_predict(request: self.request_type, background_tasks: Backgroun
...
+        if not self.stream:
+            self.app.add_middleware(GZipMiddleware, minimum_size=1000)
+        if self.max_payload_size is not None:
+            self.app.add_middleware(MaxSizeMiddleware, max_size=self.max_payload_size)
+        if self.middlewares is not None:
+            for middleware, kwargs in self.middlewares:
+                self.app.add_middleware(middleware, **kwargs)
+        if self.exc_handlers is not None:
+            for status_code, exc_handler in self.exc_handlers:
+                self.app.add_exception_handler(status_code, exc_handler)
```
Comment (aniketmaurya) [inline src/litserve/server.py:683]:
```suggestion
```

Comment (aniketmaurya) [reply src/litserve/server.py:683]: pls keep it minimal

File: tests/e2e/test_e2e.py [hunk 5]
```diff
@@ -80,6 +80,21 @@ def test_e2e_default_batching():
...
 
 
+@e2e_from_file("tests/e2e/default_exc_handler.py")
+def test_e2e_default_exc_handler():
+    resp = requests.post("http://127.0.0.1:8000/predict", json={"base64": "_9j_4AAQSkZJRgABAQEAYAB"}, headers=None)
+    assert resp.status_code == 400, f"Expected response to be 400 but got {resp.status_code}"
+    assert resp.json() == {"message": "cannot decode input"}, "server didn't return expected output"
+
+
+@e2e_from_file("tests/e2e/default_middleware.py")
```
Comment (aniketmaurya) [inline tests/e2e/test_e2e.py:90]: Please create a dummy middleware, pass that to LitServer and assert the test.

**Review by aniketmaurya — CHANGES_REQUESTED** [review]
Awesome first PR 🚀 ! Added few comments. Please let me know if you need any help.

**Comment by lorenzomassimiani** [comment]
> awesome PR @lorenzomassimiani! can you add the motivation from the user perspective at the very top? "As a user we need X because of Y"
> 
> * what problem does it solve
> * etc...

Thank you!

File: src/litserve/server.py (hunk 1 above)
Comment (lorenzomassimiani) [reply src/litserve/server.py]: ok, I removed the exception handlers part.

File: src/litserve/server.py (hunk 3 above)
Comment (lorenzomassimiani) [reply src/litserve/server.py:457]: done

File: tests/e2e/test_e2e.py (hunk 5 above)
Comment (lorenzomassimiani) [reply tests/e2e/test_e2e.py:83]: Do you mean I should launch a LitServer with some dummy middleware that does nothing and add a test that checks if it responds without errors?

Comment (aniketmaurya) [reply tests/e2e/test_e2e.py:83]: yes, maybe the dummy middleware can modify the request data and you can assert it in LitAPI.decode_method. 

Example:

* the client input is `{"input": 4.0}`
* Middleware changes it to `{"input": 5.0}`
* LitAPI.decode_method `assert request["input"]==5.0`

Comment (aniketmaurya) [reply tests/e2e/test_e2e.py:83]: I think we can move the class definition to `tests/test_lit_server.py` and then delete `tests/e2e/default_middleware.py`. You can write a test similar to [this](https://github.com/Lightning-AI/LitServe/blob/74d80e78af12411e05054da7a1b62fefb8135106/tests/test_lit_server.py#L454).

Comment (lorenzomassimiani) [reply tests/e2e/test_e2e.py]: ok, I moved the test from e2e to test_lit_server.

**Comment by williamFalcon** [comment]
>As a user I need to enrich the litserve server with fastapi middleware, for example to update API metrics.

thank you! love this. Can you expand (mostly for my own knowledge), what is an example of "update API metrics", ie: a tangible use case. 

i don't disagree this feature is important, but i want to learn more about what the lack of the feature is blocking for real use cases.

**Comment by lorenzomassimiani** [comment]
> > As a user I need to enrich the litserve server with fastapi middleware, for example to update API metrics.
> 
> thank you! love this. Can you expand (mostly for my own knowledge), what is an example of "update API metrics", ie: a tangible use case.
> 
> i don't disagree this feature is important and we should land it. But on a side note, i want to learn more about what the lack of this feature is blocking for real use cases.

Yes, we use it to send useful API statistics, such as response time and status, to Graphite to monitor the application in production (but this is only an example on how to use middleware in a fastapi app). It's useful to use a middleware in this cases because this operations are executed on top of the api call.

**Comment by williamFalcon** [comment]
got it... so it's conceptually the same that we achieve with lifecycle callbacks in PyTorch Lightning (https://lightning.ai/docs/pytorch/stable/extensions/callbacks.html).

Would a callback system be more useful here instead of middleware? (that still allows using FastAPI middlewares if you have them).

**Comment by lorenzomassimiani** [comment]
@williamFalcon yes, it could be an other interesting way to do it. We use a similar approach for updating single function elapsing time by decorating them with a function that executes them and then send the execution time to graphite. But for the stats that are about the entire api call we prefered to use a middleware that is executed after it.

File: tests/test_lit_server.py [hunk 6]
```diff
@@ -457,3 +459,24 @@ def test_http_exception():
...
+        return response
+
+
+def test_custom_middleware():
+    server = ls.LitServer(ls.examples.SimpleLitAPI(), middlewares=[(RequestIdMiddleware, {"length": 5})])
+    with wrap_litserve_start(server) as server, TestClient(server.app) as client:
+        response = client.post("/predict", json={"input": 4.0})
+        assert response.status_code == 200, f"Expected response to be 200 but got {response.status_code}"
+        assert response.json() == {"output": 16.0}, "server didn't return expected output"
+        assert response.headers["X-Request-Id"] == "00000"
```
Comment (aniketmaurya) [inline tests/test_lit_server.py:482]: love this 🔥

**Comment by lorenzomassimiani** [comment]
I found that ci testing failed due an import error from starlette (probably caused by version mismatching) https://github.com/Lightning-AI/LitServe/actions/runs/10618104701/job/29434334978?pr=241#step:8:20 .
What do you think about using a more generic `Optional[list[tuple[type, dict]]]`?

**Comment by williamFalcon** [comment]
ok. let’s add callbacks as a follow up @aniketmaurya

File: src/litserve/server.py [hunk 7]
```diff
@@ -422,13 +423,16 @@ def __init__(
         stream: bool = False,
         spec: Optional[LitSpec] = None,
         max_payload_size=None,
+        middlewares: Optional[list[tuple[type[_MiddlewareClass[P]]], dict]] = None,
```
Comment (aniketmaurya) [inline src/litserve/server.py:426]:
```suggestion
        middlewares: Optional[List[Tuple[Callable, dict]]] = None,
```

File: src/litserve/server.py [hunk 8]
```diff
@@ -36,6 +36,7 @@
 from fastapi.security import APIKeyHeader
 from starlette.formparsers import MultiPartParser
 from starlette.middleware.gzip import GZipMiddleware
+from starlette.middleware import _MiddlewareClass, P
```
Comment (aniketmaurya) [inline src/litserve/server.py:39]:
```suggestion
from typing import Callable
```

**Review by aniketmaurya — APPROVED** [review]
Looks great @lorenzomassimiani!! Last two things to take care of ^^

**Comment by lorenzomassimiani** [comment]
Perfect @aniketmaurya, I just pushed the final changes.

**Comment by aniketmaurya** [comment]
thanks @lorenzomassimiani!! there are some conflicts to resolve before we can merge. Can you please resolve it? I am not able to edit the PR from my side to fix that.

**Comment by lorenzomassimiani** [comment]
done @aniketmaurya

**Pull request merged**